[nanii] 1주차 NEWS STAND 학습 계획서

bread1022 opened the pull request:
# NANII - FE2. NEWS STAND

## 학습계획서

## 📌 학습 키워드
- [x] grid
- [x] auto rolling
- [x] crawling

## 🔎 기능 요구 사항
- [x] 그리드 UI (HTML, CSS)
- [x] 상단 자동 롤링 UX
- [ ] 콘텐츠 좌우 이동

## 💥 TEST CASE
1. **기본 화면 구성**
     - [x] 기본상단역역 : 로고, 시스템시간
     - [x] 자동롤링영역 : 최신뉴스 왼쪽바, 오른쪽바
     - [x] 기사영역 : 전체언론사, 그리드보기 (기본값)
2. **최신 뉴스 자동 롤링 영역**
     - [ ] 왼쪽, 오른쪽 헤드라인 바에 각각 다른 5개의 뉴스 헤드라인을 띄워야한다
     - [x] 5초마다 자동으로 헤드라인이 위로 올라가는 애니메이션을 구현해야한다
     - [ ] 좌우 영역이 동시에 롤링되지않고 1초의 시간차로 롤링되어야한다
     - [x] 헤드라인 바 영역에 마우스 호버시, 자동 롤링을 멈추고 해당 헤드라인에 밑줄을 표시해야한다
3. **전체 언론사**
     - [ ] 기본값은 **그리드** 상태여야한다
     - [ ] 페이지 새로 고침시, 랜덤으로 언론사별 브랜드마크를 중앙으로 배치해야한다
     - [ ] 그리드 좌우 화살표로 페이지 넘김이 가능해야한다
       - [ ] 가장 첫 페이지, 마지막 페이지는 화살표 이전, 이후 화살표를 표시하지 않아야한다
     - [ ] 브랜드마크 셀에 마우스 호버시, [구독하기] 버튼이 보여야한다
       - [ ] 구독하기 버튼 클릭시, 스낵바 5초간 유지해야한다 ("내가 구독한 언론사에 추가되었습니다")
       - [ ] 5초 후, 내가 구독한 언론사 탭 리스트 보기 화면으로 즉시 이동해야한다

## 📝 주요 개발 구현 내용
[Notion-link](https://bebetter-nanii.notion.site/News-stand-f366288cc7304a9ca71dfefae5a7155e)

## 🤔 고민한 점, 느낀점
- 컴포넌트단위로 쪼개고, 새로운 패턴을 적용하려니 템플릿을 만들고, 이벤트는 어디서 적용하고? 그 이벤트는 어디서 받아와야할지 역할을 구분하는 게 너무 헷갈린다.

bread1022 committed:
docs: update README.md

bread1022 committed:
chore: 프로젝트 초기 환경 설정

hoongding commented on public/index.html:
```diff
@@ -0,0 +1,12 @@
+  <title>New Stand</title>
+</head>
+<body>
+  <div id="root"></div>
```
동적으로 요소들을 생성하겠다는 HTML 처럼 보여서 벌써 두근거리네요..!

hoongding commented on readme.md:
```diff
@@ -1 +1,56 @@
+- [ ] 상단 자동 롤링 UX
+- [ ] 콘텐츠 좌우 이동
+
+## 💥 TEST CASE
+1. **기본 화면 구성**
+     - [ ] 기본상단역역 : 로고, 시스템시간
+     - [ ] 자동롤링영역 : 최신뉴스 왼쪽바, 오른쪽바
+     - [ ] 기사영역 : 전체언론사, 그리드보기 (기본값)
+2. **최신 뉴스 자동 롤링 영역**
+     - [ ] 왼쪽, 오른쪽 헤드라인 바에 각각 다른 5개의 뉴스 헤드라인을 띄워야한다
+     - [ ] 5초마다 자동으로 헤드라인이 위로 올라가는 애니메이션을 구현해야한다
+     - [ ] 좌우 영역이 동시에 롤링되지않고 1초의 시간차로 롤링되어야한다
+     - [ ] 헤드라인 바 영역에 마우스 호버시, 자동 롤링을 멈추고 해당 헤드라인에 밑줄을 표시해야한다
+3. **전체 언론사**
+     - [ ] 기본값은 **그리드** 상태여야한다
+     - [ ] 페이지 새로 고침시, 랜덤으로 언론사별 브랜드마크를 중앙으로 배치해야한다
+     - [ ] 그리드 좌우 화살표로 페이지 넘김이 가능해야한다
+       - [ ] 가장 첫 페이지, 마지막 페이지는 화살표 이전, 이후 화살표를 표시하지 않아야한다
+     - [ ] 브랜드마크 셀에 마우스 호버시, [구독하기] 버튼이 보여야한다
+       - [ ] 구독하기 버튼 클릭시, 스낵바 5초간 유지해야한다 ("내가 구독한 언론사에 추가되었습니다")
+       - [ ] 5초 후, 내가 구독한 언론사 탭 리스트 보기 화면으로 즉시 이동해야한다
+4. 전체 언론사 **리스트 보기**
+     - [ ] 기본 화면 구성
+       - [ ] 각 분야 7 카테고리 (종합/경제, 방송/통신, IT, 영자지, 스포츠/연애, 매거진/전문지, 지역)
+       - [ ] 언론사 마크, 최종 편집일시, 구독하기 버튼, 리스트/그리드 버튼
+       - [ ] 메인 뉴스 썸네일, 타이틀
+       - [ ] 서브 뉴스 타이틀
+       - [ ] 편집권 안내 문구
+     - [ ] 선택된 카테고리의 경우 해당 카테고리의 배경이 progress bar가 되어 1초마다 색이 차오르게 애니메이션을 적용해야하며(총 20초) 해당 카테고리 전체 길이 중 현재 순서를 표시해야한다 (순서는 새로고침때마다 랜덤으로 바뀐다)
+     - [ ] 한 언론사당 20초간 화면을 보여주고 해당 카테고리에 해당하는 언론사를 다 돌면 다음 카테고리로 이동해야한다
+     - [ ] 모든 카테고리를 다 돌면 다시 처음 카테고리, 첫 언론사로 이동해야한다
+     - [ ] 좌우화살표 클릭시, 언론사 이동이 가능해야한다
+     - [ ] 선택된 카테고리 외 다른 카테고리 마우스 호버시, 밑줄 표시해야한다
+     - [ ] 선택된 카테고리 외 다른 카테고리 마우스 클릭시, 해당 페이지로 이동해야한다
+     - [ ] 메인 뉴스 썸네일 마우스 호버시, 썸네일을 5% 확대해야하고, 타이틀은 밑줄 표시해야한다
+     - [ ] 서브메뉴 마우스 호버시, 타이틀은 밑줄 표시해야한다
```
테스트 케이스가 명확하게 나눠져 있어서 저도 나니의 테스트 케이스 좀 참고하면서 고쳐봐야겠어요...
수고하셨습니다~

bread1022 committed:
feat: 기본 화면 정적으로 구현

- html 구조 잡고
- sass 적용함

bread1022 committed:
feat: header element 시간 옵션 받아서 root에 append

- 시스템 시간 옵션 :  { year: "numeric", month: "numeric", day: "numeric", weekday: "long" }

bread1022 committed:
style: Icon hover, 색상 변경하는 함수 적용

bread1022 committed:
feat: RollingBar class 구현

- template 생성한 뒤
- requestAnimationFrame 적용하여 자동 롤링 UX 구현
- raf State 이용하여 mouseover 했을 경우 롤링 멈춤 & 밑줄 구현
(애니메이션 우선적용. 아직 메서드 분리, 데이터 Fetch 적용 필요함)

bread1022 committed:
feat: main 태그 분류버튼 생성

- sortButtons 템플릿리터럴 생성 (이벤트x)
- grid 정적화면에 scss 적용

crongro commented on .gitignore:
```diff
@@ -0,0 +1 @@
+node_modules/
```
.gitignore 자동으로 생성해주던데, 직접 만드셨네요.
참고로. 어떤 것이 들어가면 좋을지 좀더 알아보세요

crongro commented on package.json:
```diff
@@ -0,0 +1,30 @@
+  "description": "## 학습계획서",
+  "main": "index.js",
+  "type": "module",
+  "scripts": {
```
npm script 잘 써서 좋네요.

crongro commented on public/index.html:
```diff
@@ -0,0 +1,93 @@
+      <div class="view-container">
+        <div class="grid-container">
+          <ul class="grid-area">
+            <li><img src="https://s.pstatic.net/static/newsstand/2020/logo/light/0604/973.png"></li>
+            <li><img src="https://s.pstatic.net/static/newsstand/2020/logo/light/0604/973.png"></li>
+            <li><img></li>
+            <li><img></li>
+            <li><img></li>
+            <li><img></li>
+            <li><img></li>
+            <li><img></li>
```
가급적 html 에 반복적인 리스트형태의 태그표현은 자제하고, 
 js로 동적으로 추가하는걸로~

crongro commented on src/sass/layout/_sortButtons.scss:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  .view-buttons {
+    @extend .press-buttons;
```
include, extend 등 잘 활용했네요.

crongro commented on src/js/components/section/rollingBar.js:
```diff
@@ -0,0 +1,80 @@
+
+class RollingBar {
+  #PANEL_COUNT = 2;
+  #TRANSLATE_TIME = '500ms';
```
이것도 생성자에서 옵션으로 받을수도 있죠~ 
default parameter 활용도 하시고요.

crongro commented on src/js/components/section/rollingBar.js:
```diff
@@ -0,0 +1,80 @@
+import createEl from '../../utils/util.js';
+
+class RollingBar {
```
클래스 잘 구현했네요. 동작만 잘 되면 아주 좋을듯~ (잘되죠?)

crongro commented on src/js/components/section/rollingBar.js:
```diff
@@ -0,0 +1,80 @@
+  init() {
+    this.rollingBar.innerHTML = this.template();
+    this.rollingPanelHandler();
+    return this.rollingBar;
```
보통 초기화 하는 함수는 반환값이 없긴하죠~ 이름도 init 이기도 하고요.

crongro commented on src/sass/layout/_section.scss:
```diff
@@ -0,0 +1,47 @@
+        height: 16px;
+        will-change: transform;
+
+        .flick-panel {
```
sass로 구현시, 중첩된 클래스가 너무 깊지 않도록 하는것이 좋은데요.
지금정도까지는 괜찮은정도 입니다.

crongro commented on src/js/app.js:
```diff
@@ -0,0 +1,17 @@
+  const root = document.querySelector("#root");
+
+  root.append(header(CONSTANTS["NEWS_STAND"], systemTimeOption));
+  root.append(new RollingBar(CONSTANTS["ROLLING_LINK_PRESS"], 5000, 'left', 'right').init());
```
한 줄로 표현해도 좋긴하고요.
이부분을 나눠서 변수로 담아두는것도 좋고요.
new RollingBar(CONSTANTS["ROLLING_LINK_PRESS"], 5000, 'left', 'right').init()

crongro reviewed (approved):
수고하셨어요.
css코드 잘하셨고, 롤링부터 먼저 시작한 것도 좋군요.
